File: servers/finance/pkg/errors/api_error.go
package errors

import (
	"encoding/json"
	"net/http"
	"time"
)

// ErrorCode represents a machine-readable error code
type ErrorCode string

// Standard error codes used across all services
const (
	ErrCodeValidationError    ErrorCode = "VALIDATION_ERROR"
	ErrCodeNotFound           ErrorCode = "NOT_FOUND"
	ErrCodeUnauthorized       ErrorCode = "UNAUTHORIZED"
	ErrCodeForbidden          ErrorCode = "FORBIDDEN"
	ErrCodeConflict           ErrorCode = "CONFLICT"
	ErrCodeInternalError      ErrorCode = "INTERNAL_ERROR"
	ErrCodeRateLimited        ErrorCode = "RATE_LIMITED"
	ErrCodeServiceUnavailable ErrorCode = "SERVICE_UNAVAILABLE"
	ErrCodeBadRequest         ErrorCode = "BAD_REQUEST"
)

// ErrorDetail contains the error information
type ErrorDetail struct {
	Code    ErrorCode   `json:"code"`
	Message string      `json:"message"`
	Details interface{} `json:"details,omitempty"`
}

// MetaInfo contains request metadata
type MetaInfo struct {
	RequestID string `json:"requestId"`
	TraceID   string `json:"traceId,omitempty"`
	Timestamp string `json:"timestamp"`
	Service   string `json:"service"`
}

// ApiErrorResponse is the unified error response format
type ApiErrorResponse struct {
	Error ErrorDetail `json:"error"`
	Meta  MetaInfo    `json:"meta"`
}

// ApiError represents an API error with code and HTTP status
type ApiError struct {
	Code       ErrorCode
	StatusCode int
	Message    string
	Details    interface{}
}

// Error implements the error interface
func (e *ApiError) Error() string {
	return string(e.Code) + ": " + e.Message
}

// ToResponse converts the error to an ApiErrorResponse
func (e *ApiError) ToResponse(requestID, service, traceID string) ApiErrorResponse {
	return ApiErrorResponse{
		Error: ErrorDetail{
			Code:    e.Code,
			Message: e.Message,
			Details: e.Details,
		},
		Meta: MetaInfo{
			RequestID: requestID,
			TraceID:   traceID,
			Timestamp: time.Now().UTC().Format(time.RFC3339),
			Service:   service,
		},
	}
}

// WriteErrorResponse writes an ApiErrorResponse to the http.ResponseWriter
func (e *ApiError) WriteErrorResponse(w http.ResponseWriter, requestID, service, traceID string) {
	response := e.ToResponse(requestID, service, traceID)
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(e.StatusCode)
	json.NewEncoder(w).Encode(response)
}

// ErrorCodeToHTTPStatus maps error codes to HTTP status codes
func ErrorCodeToHTTPStatus(code ErrorCode) int {
	switch code {
	case ErrCodeValidationError, ErrCodeBadRequest:
		return http.StatusBadRequest
	case ErrCodeUnauthorized:
		return http.StatusUnauthorized
	case ErrCodeForbidden:
		return http.StatusForbidden
	case ErrCodeNotFound:
		return http.StatusNotFound
	case ErrCodeConflict:
		return http.StatusConflict
	case ErrCodeRateLimited:
		return http.StatusTooManyRequests
	case ErrCodeServiceUnavailable:
		return http.StatusServiceUnavailable
	case ErrCodeInternalError:
		return http.StatusInternalServerError
	default:
		return http.StatusInternalServerError
	}
}

// New creates a new ApiError
func New(code ErrorCode, message string, details interface{}) *ApiError {
	return &ApiError{
		Code:       code,
		StatusCode: ErrorCodeToHTTPStatus(code),
		Message:    message,
		Details:    details,
	}
}

// NotFound creates a NOT_FOUND error
func NotFound(resource string) *ApiError {
	return New(ErrCodeNotFound, resource+" not found", nil)
}

// NotFoundWithID creates a NOT_FOUND error with ID
func NotFoundWithID(resource, id string) *ApiError {
	return New(ErrCodeNotFound, resource+" with id '"+id+"' not found", nil)
}

// Validation creates a VALIDATION_ERROR
func Validation(message string, details interface{}) *ApiError {
	return New(ErrCodeValidationError, message, details)
}

// Unauthorized creates an UNAUTHORIZED error
func Unauthorized(message string) *ApiError {
	if message == "" {
		message = "Authentication required"
	}
	return New(ErrCodeUnauthorized, message, nil)
}

// Forbidden creates a FORBIDDEN error
func Forbidden(message string) *ApiError {
	if message == "" {
		message = "Access denied"
	}
	return New(ErrCodeForbidden, message, nil)
}

// Internal creates an INTERNAL_ERROR
func Internal(message string) *ApiError {
	if message == "" {
		message = "An internal error occurred"
	}
	return New(ErrCodeInternalError, message, nil)
}

// RateLimited creates a RATE_LIMITED error
func RateLimited(message string) *ApiError {
	if message == "" {
		message = "Too many requests"
	}
	return New(ErrCodeRateLimited, message, nil)
}

// ServiceUnavailable creates a SERVICE_UNAVAILABLE error
func ServiceUnavailable(message string) *ApiError {
	if message == "" {
		message = "Service temporarily unavailable"
	}
	return New(ErrCodeServiceUnavailable, message, nil)
}

// Conflict creates a CONFLICT error
func Conflict(message string) *ApiError {
	return New(ErrCodeConflict, message, nil)
}

// FromError converts any error to ApiError
func FromError(err error) *ApiError {
	if err == nil {
		return nil
	}

	if apiErr, ok := err.(*ApiError); ok {
		return apiErr
	}

	return Internal(err.Error())
}

// WriteResponse is a helper function to write an error response
func WriteResponse(w http.ResponseWriter, code ErrorCode, message, requestID, service, traceID string) {
	apiErr := New(code, message, nil)
	apiErr.WriteErrorResponse(w, requestID, service, traceID)
}

File: packages/common/src/errors/api-error.ts
/**
 * Unified API Error Response Format
 * Used across all services for consistent error handling
 */

export interface ApiErrorResponse {
  error: {
    code: string;
    message: string;
    details?: unknown;
  };
  meta: {
    requestId: string;
    traceId?: string;
    timestamp: string;
    service: string;
  };
}

/**
 * Standard error codes used across all services
 */
export const ErrorCodes = {
  VALIDATION_ERROR: "VALIDATION_ERROR",
  NOT_FOUND: "NOT_FOUND",
  UNAUTHORIZED: "UNAUTHORIZED",
  FORBIDDEN: "FORBIDDEN",
  CONFLICT: "CONFLICT",
  INTERNAL_ERROR: "INTERNAL_ERROR",
  RATE_LIMITED: "RATE_LIMITED",
  SERVICE_UNAVAILABLE: "SERVICE_UNAVAILABLE",
  BAD_REQUEST: "BAD_REQUEST",
} as const;

export type ErrorCode = (typeof ErrorCodes)[keyof typeof ErrorCodes];

/**
 * HTTP status codes mapped to error codes
 */
export const ErrorCodeToHttpStatus: Record<ErrorCode, number> = {
  [ErrorCodes.VALIDATION_ERROR]: 400,
  [ErrorCodes.BAD_REQUEST]: 400,
  [ErrorCodes.UNAUTHORIZED]: 401,
  [ErrorCodes.FORBIDDEN]: 403,
  [ErrorCodes.NOT_FOUND]: 404,
  [ErrorCodes.CONFLICT]: 409,
  [ErrorCodes.RATE_LIMITED]: 429,
  [ErrorCodes.INTERNAL_ERROR]: 500,
  [ErrorCodes.SERVICE_UNAVAILABLE]: 503,
};

/**
 * Custom API Error class for consistent error handling
 */
export class ApiError extends Error {
  public readonly code: ErrorCode;
  public readonly statusCode: number;
  public readonly details?: unknown;

  constructor(code: ErrorCode, message: string, details?: unknown) {
    super(message);
    this.name = "ApiError";
    this.code = code;
    this.statusCode = ErrorCodeToHttpStatus[code];
    this.details = details;
  }

  /**
   * Convert error to API response format
   */
  toResponse(
    requestId: string,
    service: string,
    traceId?: string,
  ): ApiErrorResponse {
    return {
      error: {
        code: this.code,
        message: this.message,
        details: this.details,
      },
      meta: {
        requestId,
        traceId,
        timestamp: new Date().toISOString(),
        service,
      },
    };
  }

  /**
   * Create a NOT_FOUND error
   */
  static notFound(resource: string, id?: string): ApiError {
    const message = id
      ? `${resource} with id '${id}' not found`
      : `${resource} not found`;
    return new ApiError(ErrorCodes.NOT_FOUND, message);
  }

  /**
   * Create a VALIDATION_ERROR
   */
  static validation(message: string, details?: unknown): ApiError {
    return new ApiError(ErrorCodes.VALIDATION_ERROR, message, details);
  }

  /**
   * Create an UNAUTHORIZED error
   */
  static unauthorized(message = "Authentication required"): ApiError {
    return new ApiError(ErrorCodes.UNAUTHORIZED, message);
  }

  /**
   * Create a FORBIDDEN error
   */
  static forbidden(message = "Access denied"): ApiError {
    return new ApiError(ErrorCodes.FORBIDDEN, message);
  }

  /**
   * Create an INTERNAL_ERROR
   */
  static internal(message = "An internal error occurred"): ApiError {
    return new ApiError(ErrorCodes.INTERNAL_ERROR, message);
  }

  /**
   * Create a RATE_LIMITED error
   */
  static rateLimited(message = "Too many requests"): ApiError {
    return new ApiError(ErrorCodes.RATE_LIMITED, message);
  }

  /**
   * Create a SERVICE_UNAVAILABLE error
   */
  static serviceUnavailable(
    message = "Service temporarily unavailable",
  ): ApiError {
    return new ApiError(ErrorCodes.SERVICE_UNAVAILABLE, message);
  }

  /**
   * Create a CONFLICT error
   */
  static conflict(message: string): ApiError {
    return new ApiError(ErrorCodes.CONFLICT, message);
  }

  /**
   * Check if an error is an ApiError
   */
  static isApiError(error: unknown): error is ApiError {
    return error instanceof ApiError;
  }

  /**
   * Convert any error to ApiError
   */
  static from(error: unknown): ApiError {
    if (ApiError.isApiError(error)) {
      return error;
    }

    if (error instanceof Error) {
      return new ApiError(ErrorCodes.INTERNAL_ERROR, error.message);
    }

    return new ApiError(
      ErrorCodes.INTERNAL_ERROR,
      "An unexpected error occurred",
    );
  }
}

/**
 * Build an error response object
 */
export function buildErrorResponse(
  error: ApiError | Error | unknown,
  requestId: string,
  service: string,
  traceId?: string,
): { response: ApiErrorResponse; statusCode: number } {
  const apiError = ApiError.from(error);
  return {
    response: apiError.toResponse(requestId, service, traceId),
    statusCode: apiError.statusCode,
  };
}
